### ISSUE TITLE
t.w.xmlrpc.QueryProtocol does not close the connection after handling the request

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2518]
|<img alt="karlb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| karlb reported|
|-|-|
|Trac ID|trac#2518|
|Type|defect|
|Created|2007-03-12 16:03:02Z|
|Branch|https://github.com/twisted/twisted/tree/close-connection-after-request-2518|

When using an xmlrpc.Proxy instance in a test, trial often complains about left over twisted.internet.tcp.Client instances.

Changing QueryProtocol.handleResponse from
```python
    def handleResponse(self, contents):
        self.factory.parseResponse(contents)
```
to
```python
    def handleResponse(self, contents):
        self.factory.parseResponse(contents)
        self.transport.loseConnection()
```
resolves the problem. This seems to make sense and I can't see any disadvantages.

<details><summary>Searchable metadata</summary>

```
trac-id__2518 2518
type__defect defect
reporter__karlb karlb
priority__normal normal
milestone__ 
branch__branches_close_connection_after_request_2518 branches/close-connection-after-request-2518
branch_author__jml jml
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1173715382000000 1173715382000000
changetime__1289401002000000 1289401002000000
version__None None
owner__ 
cc__karlb
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|

(In [30183]) Branching to 'close-connection-after-request-2518'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml commented|
|-|-|

I just sat down to try to make this change and couldn't figure out how to write the tests for it.

Specifically, I can't think of a good way to get at the transport to see if it is connected.

Any thoughts?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Shouldn't trial fail the test for leaving the connection open?

Aside from that, though, hook in `queryFactory`, which lets you hook in to the protocol, which gives you the transport.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml commented|
|-|-|

It's intermittent. I'll hook in as you suggest.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

> It's intermittent.

If it talked to a server that didn't close the connection itself, it would probably become reliable.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml commented|
|-|-|

Yeah, but I'm not clear how to do that.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_8"></a>|@jml commented|
|-|-|

(In [30184]) Disconnect from the server before handling the response.
refs [#2518](https://github.com/twisted/twisted/issues/2518)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @jml|
|-|-|

  1. Needs news fragment
  1. Needs docstring on `handleResponse`

One thing that occurs to me is that this still won't cause the transport to be closed by the time the result Deferred fires.  Instead it will probably take another iteration for `disconnecting` to actually turn into `disconnected`.  However, this is good enough to make unit tests happy (which I think is the goal?) since trial will iterate the reactor a couple times after each test.

So, please merge when the above two points are fixed.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_11"></a>|@jml commented|
|-|-|

Thanks very much for the review.

I noticed that an iteration is needed after the `Deferred` fires in some later testing.  The goal is, indeed, to make the unit tests happy, so this patch is good enough. However, do you know of a way we could avoid the need for an extra iteration?

(Adding the news fragment and docstring now)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_12"></a>|@jml commented|
|-|-|

(In [30254]) Docstring and NEWS fragment (refs [#2518](https://github.com/twisted/twisted/issues/2518))

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_13"></a>|@jml removed owner|
|-|-|

I've changed the implementation approach to avoid the need for iteration (I can't believe I didn't figure this out for myself). It's different enough that it needs a re-review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set owner to @jml|
|-|-|

 * 
```
+        if self._contents:
```

I would do a "if is not None" check instead, to handle empty results properly.

 * self._contents would probably be better named self._response, no?

Please merge once fixed.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_15"></a>|@jml commented|
|-|-|

(In [30261]) Rename 'contents' to 'response', and be explicit about checking None
refs [#2518](https://github.com/twisted/twisted/issues/2518)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_16"></a>|@jml set status to `closed`|
|-|-|

(In [30262]) Merge close-connection-after-request-2518

 * Author: jml
 * Reviewers: exarkun, therve
 * Fixes [#2518](https://github.com/twisted/twisted/issues/2518)

Make sure that the Deferred returned by our XML-RPC client
only fires after the connection has been closed.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
